src/lib.rs

```diff
@@ -0,0 +1,780 @@
+    /// because the way in which we spill a real register may depend on the
+    /// value that we are using it for. E.g., if a machine has V128 registers
+    /// but we also use them for F32 and F64 values, we may use a different
+    /// store-slot size and smaller-operand store/load instructions for an F64
+    /// than for a true V128.
+    fn spillslot_size(&self, regclass: RegClass, for_vreg: VReg) -> usize;
+
+    /// When providing a spillslot number for a multi-slot spillslot,
+    /// do we provide the first or the last? This is usually related
+    /// to which direction the stack grows and different clients may
+    /// have different preferences.
+    fn multi_spillslot_named_by_last_slot(&self) -> bool {
+        false
+    }
+}
+
+/// A position before or after an instruction.
+#[derive(Clone, Copy, Debug, PartialEq, Eq, PartialOrd, Ord, Hash)]
+#[repr(u8)]
+pub enum InstPosition {
+    Before = 0,
+    After = 1,
+}
+
+/// A program point: a single point before or after a given instruction.
+#[derive(Clone, Copy, Debug, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub struct ProgPoint {
+    pub inst: Inst,
+    pub pos: InstPosition,
+}
+
+impl ProgPoint {
+    pub fn before(inst: Inst) -> Self {
+        Self {
+            inst,
+            pos: InstPosition::Before,
+        }
+    }
+
+    pub fn after(inst: Inst) -> Self {
+        Self {
+            inst,
+            pos: InstPosition::After,
+        }
+    }
+
+    pub fn next(self) -> ProgPoint {
+        match self.pos {
+            InstPosition::Before => ProgPoint {
+                inst: self.inst,
+                pos: InstPosition::After,
+            },
+            InstPosition::After => ProgPoint {
+                inst: self.inst.next(),
+                pos: InstPosition::Before,
+            },
+        }
+    }
+
+    pub fn prev(self) -> ProgPoint {
+        match self.pos {
+            InstPosition::Before => ProgPoint {
+                inst: self.inst.prev(),
+                pos: InstPosition::After,
+            },
+            InstPosition::After => ProgPoint {
+                inst: self.inst,
+                pos: InstPosition::Before,
+            },
+        }
+    }
+
+    pub fn to_index(self) -> u32 {
+        debug_assert!(self.inst.index() <= ((1 << 31) - 1));
+        ((self.inst.index() as u32) << 1) | (self.pos as u8 as u32)
+    }
+
+    pub fn from_index(index: u32) -> Self {
+        let inst = Inst::new((index >> 1) as usize);
+        let pos = match index & 1 {
+            0 => InstPosition::Before,
+            1 => InstPosition::After,
+            _ => unreachable!(),
+        };
+        Self { inst, pos }
+    }
+}
+
+/// An instruction to insert into the program to perform some data movement.
+#[derive(Clone, Debug)]
+pub enum Edit {
+    /// Move one allocation to another. Each allocation may be a
+    /// register or a stack slot (spillslot).
+    Move { from: Allocation, to: Allocation },
```

Comment:
Does the allocator generate stack-to-stack moves? Does the caller need to be able to handle this case?

Reply:
Yes, it does at the moment; I haven't decided whether it would be better to require the client to handle this (e.g. with another scratch register, or maybe a push-mem/pop-mem sequence on x86) or add some special logic here. I suppose `ParallelMoves` could have an option to use a scratch register to move between stack slots, though I think in some cases that would require two scratch registers to be set aside (e.g. `{slot1, slot0} := {slot0, slot1}`).

Reply:
Fixed this since the last comment -- the allocator will now never generate stack-to-stack moves.

Reply:
(And added doc-comment just now to note this.)